### ISSUE TITLE
Document regular expression font mapping

### ISSUE THREAD

**Issue by Aceto1** [jstockwin/py-pdf-parser#237]
I would like it to be possible to filter elements by parts of font names. E.g. .filter_by_font_contains("Arial,Bold,12.0")

I have a few hundred PDFs that have a very similar structure but for some reason the fonts have different prefixes even when used for the same elements across PDFs.
E.g.:
* CIBPIK+Arial,Bold,12.0
* JHCMFI+Arial,Bold,12.0
* JHIENK+Arial,Bold,12.0

Each one is the font for table headers in a PDF.

A function like .filter_by_font_contains("Arial,Bold,12.0") would enable me to just filter by that part of the font name (which is consistent for me across PDFs). Even better would be a function that takes an options object (no idea if this is done in Python) e.g.:
```
{
  fontName: "Arial",
  bold: true,
  size: 12.0
}
```
so the input isn't limited to a continuous string. Would be helpful if you were trying to filter for all elements that have Arial, 12.0 without either excluding the ones that are bold, or doing two queries and merging them later.

Since I'm very new to Python I can certainly try to help but I can't promise anything.

**Comment by jstockwin** [comment]
Hey @Aceto1, I think that's a nice idea! Particularly being able to filter for more structured font info (e.g. just the size, or just the name). I don't think I have e.g. "Bold" as an attribute, though - it just comes as a string. But I could split the "Arial,Bold" from the size, certainly.

I think in your case there is an out-of-the-box solution, though. You can pass a `font_mapping`, to map the real font names (e.g. `CIBPIK+Arial,Bold,12.0`) to your own font names (e.g. `arial_bold_12`, or even better something more meaningful like `page_header` or `bold_text`). You are able to pass the font mapping as regular expressions if you pass `font_mapping_is_regex=True`. (See the [font mapping overview](https://py-pdf-parser.readthedocs.io/en/latest/overview.html?highlight=font%20mapping#font-mappings), the [Order Summary example](https://py-pdf-parser.readthedocs.io/en/latest/examples/order_summary.html?highlight=font%20mapping#step-2-use-a-font-mapping), and the [PDFDocument reference](https://py-pdf-parser.readthedocs.io/en/latest/reference/components.html#py_pdf_parser.components.PDFDocument) in the docs).

The regular expression font mapping is not well documented, this could be improved. In particular, the order summary example has a similar issue (but the weird prefixes don't change for each font name), so we could easily point this out there.

In your case, then, perhaps something similar to the following would help solve your issue:
```
_FONT_MAPPING = {
    r"\w{6}\+Arial,Bold,12.0": "bold_text",  # (or other sensible names)
   # Add more fonts here as needed.
}
document = load_file(..., font_mapping=_FONT_MAPPING, font_mapping_is_regex=True)

document.elements.filter_by_font("bold_text")
```
You can use whatever regular expression you like, the `\w{6}\+` will match any 6 letters and a `+` at the start of the font.

Let me know if that helps you and we can work out how to move forward with this issue! If anything is unclear, just ask!

**Comment by Aceto1** [comment]
Hey, thanks for the quick answer. I tested it and it works like a charm.
I could provide a PR for the docs if you want.

**Comment by jstockwin** [comment]
Glad it worked. Sure, a documentation PR would be fantastic!

I would personally do the following:
1. Add a note to the [Font Mapping Overview](https://py-pdf-parser.readthedocs.io/en/latest/overview.html#font-mappings), just mentioning that regular expression mapping is possible.
2. Add a new "Step 3" to the [Order Summary Example](https://py-pdf-parser.readthedocs.io/en/latest/examples/order_summary.html) which basically does step 2 again but now with regular expressions, making the comment that for some PDFs the first characters can change. Note that (a) there is a "full code" section at the bottom of each example, and (b) there are tests to ensure this continues to be correct. There will both need updating too.

Let me know if you need any help, otherwise I look forward to a PR!

**Comment by jstockwin** [comment]
Note: I've just converted this issue to be a documentation issue. Whilst other comments (about more structured font filtering) are still very valid, I don't really have the bandwidth to do them right now. If someone else is interested in that, do feel free to open a separate issue and reference this one and we can consider it again.

**Comment by jstockwin** [comment]
Hey @Aceto1, just wanted to check if you're still interested in producing a documentation PR for this?

I appreciate time is always short, but if you can let me know whether you expect to get around to it or not that would be great. Thanks!

**Comment by jstockwin** [comment]
Amazing, thank you!

**Comment by Aceto1** [comment]
Alright, is there anything else left to do on this issue that I can help with?

**Comment by jstockwin** [comment]
All good in my opinion, thanks very much!

Issue closed by #245
